### PR TITLE
Handling multipart calls in iffy

### PR DESCRIPTION
**Pull request by geoffreybauduin**
This is clearly a WIP, the Call interface needs to be rewritten to be less painful to implement.

Open to suggestions though, I'm trying to get rid of the default supposition where API only expects JSON as input

### CODE REVIEW

**Comment by loopfz** [comment]
Interesting. Indeed, we should avoid the json assumptions at some point!
I wonder if we have to put the abstraction so high up though, I feel like there's a fair amount of redundancy between JSONCall and MultipartCall. + having to define a separate AddXX function..

Could we go for some composition maybe? E.g. "body" param being an interface instead of a plain string, with some functions attached: ContentType(), Body(), HandleResponse(raw, respObj) ?

tester.AddCall("foo", "POST", "/foo", iffy.Multipart{Sections: /* raw data */, Files: /* file paths */, ...}).Checkers(...)
tester.AddCall("bar", "POST", "/bar", iffy.JSON{`{"foo":"bar"}`}).ResponseObject(&Bar{}).Checkers(...)

Only downside would be breaking the current signature that expects a plain string. Though we could even make that an interface{} + type switch to accept plain (default to json, keep current behavior) or custom interface.

Thoughts ?

File: iffy/iffy.go [hunk 1]
```diff
@@ -23,68 +21,53 @@ type Tester struct {
...
 	}
 }
 
-func (t *Tester) AddCall(name, method, querystr, body string) *Call {
+func (t *Tester) AddCall(name, method, querystr string, body interface{}) *Call {
+	var bodyInterface Body
+	switch body.(type) {
+	case Body:
+		bodyInterface = body.(Body)
+	default:
```
Comment (nomonamo) [inline iffy/iffy.go]: what about nil body? on a simple GET, for instance

Comment (geoffreybauduin) [reply iffy/iffy.go]: Added a `NoopBody` structure to handle this case. If content-type is empty, we are not inserting the content-type inside the request headers either (keeping the compat we had from before as well)

**Comment by geoffreybauduin** [comment]
I tested the branch on my side, it does what I expect it to do.
@loopfz are you happy with this interfacing or did you see it done another way ?

**Pull request closed without merging**